pep-0544.txt

```diff
@@ -0,0 +1,949 @@
+a ``__subclasshook__()`` that provides the same semantics
+for class and instance checks as for ``collections.abc`` classes::
+
+  from typing import runtime, Protocol
+
+  class Closeable(Protocol):
+      @abstractmethod
+      def close(self):
+          ...
+
+  assert isinstance(open('some/file'), Closeable)
+
+Static type checkers will understand ``isinstance(x, Proto)`` and
+``issubclass(C, Proto)`` for protocols (as they already do
+for ``Iterable`` etc.). Static type checkers will narrow types after such
+checks by the type erased ``Proto`` (i.e. with all variables having type
+``Any`` and all methods having type ``Callable[..., Any]``).
+Note that ``isinstance(x, Proto[int])`` etc. will always fail in agreement
+with PEP 484. Examples::
+
+  from typing import Iterable, Iterator, Sequence
+
+  def process(items: Iterable[int]) -> None:
+      if isinstance(items, Iterator):
+          # 'items' have type 'Iterator[int]' here
+      elif isinstance(items, Sequence[int]):
+          # Error! Can't use 'isinstance()' with subscripted protocols
+
+
+Using Protocols in Python 3.5 and Earlier
+=========================================
+
+Variable annotation syntax was added in Python 3.6, so that the syntax
+for defining protocol variables proposed in `specification`_ section can't
+be used in earlier versions. To define these in earlier versions of Python
+one can use abstract properties::
+
+  class Foo(Protocol):
+      @abstractproperty
+      def c(self) -> int: ...
+
+      @abstractproperty
+      def c(self) -> int:   # Default value can be provided for property.
+          return 0
+
+      @property
+      def e(self) -> int:   # Note, this is not a protocol member.
+          return 0
+
+The ``typing`` module changes proposed in this PEP will be also
+backported to earlier versions via the backport currently available on PyPI.
+
+
+Runtime Implementation of Protocol Classes
+==========================================
+
+Implementation details
+----------------------
+
+The runtime implementation could be done in pure Python without any
+effects on the core interpreter and standard library except in the
+``typing`` module:
+
+* Define class ``typing.Protocol`` similar to ``typing.Generic``.
+* Implement metaclass functionality to detect whether a class is
+  a protocol or not. Add a class attribute ``__protocol__ = True``
+  if that is the case. Verify that a protocol class only has protocol
+  base classes in the MRO (except for object).
+* Makes ``isinstance()`` behave as for ``collections.abc`` by adding all
+  abstract methods and all variables from ``__annotations__`` to
+  ``__subclsshook__()``.
+* In the backported version of ``typing``, translate ``...`` class attribute
+  values to properties.
+* All structural subtyping checks will be performed by static type checkers,
+  such as ``mypy`` [mypy]_. No additional support for protocol validation will
+  be provided at runtime.
+
+
+Changes in the typing module
```

Comment:
The idea is that there should be no backward incompatibility. The classes in ``typing`` module (as well as in ``collections.abc``) already behave like protocols at runtime:
```python
class Whatever:
    def __iter__(self): return []
assert isinstance(Whatever(), Iterable)
```
as I mention below "Practically, few changes will be needed in ``typing`` since some of these classes already behave the necessary way at runtime", most of these will be changed only in typeshed.

Also, after I implement the changes proposed by Jukka, all code that passes mypy now, will still pass (unless I don't see some subtleties).